### DISCUSSION TITLE
Alternative Method to Bypass the `tx.origin` Change in OpenZeppelin Contracts: Use `_transferOwnership` in Use Hello in the Main Constructor

### DISCUSSION THREAD

**Discussion by defido** [pcaversaccio/xdeployer#18] [Show and tell]
This is OpenZeppelin (OZ) [`Ownable`](https://docs.openzeppelin.com/contracts/4.x/api/access#Ownable)-specific. If you're using your own access control setup, then feel free to ignore this.

The current process using OZ's `Ownable` in combination with the `xdeployer` is the following: the `msg.sender` of the contract creation transaction is the helper smart contract [`Create2Deployer`](https://github.com/pcaversaccio/create2deployer) with address `0x13b0D85CcB8bf860b6b79AF3029fCA081AE9beF2`. This in turn means using the standard OZ `Ownable` access control method would set the `owner` to `0x13b0D85CcB8bf860b6b79AF3029fCA081AE9beF2`. One of the workarounds would be to change the `msg.sender` to `tx.origin` in the original OZ `Ownable` contract. However, there exists a further way to resolve this problem: When using OZ contracts, simply use `_transferOwnership(your wallet address)` in the main constructor setup. This adjustment will trigger first a transfer of ownership to the helper contract and thereafter a transfer of the ownership from the helper smart contract to the wallet address you set in the main constructor (2 events will be emitted); everything within the same contract creation transaction. See also a more comprehensive discussion [here](https://github.com/pcaversaccio/xdeployer/pull/17).

**Comment by tab00** [comment]
Instead of `Ownable` it's better to use `AccessControl`:
```solidity
import "@openzeppelin/contracts/access/AccessControl.sol";
```

```solidity
contract XXX is ERC20, AccessControl {
    bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
```

Change `msg.sender` to `tx.origin` in the `_grantRole` and `_mint` calls:
```solidity
    constructor() ERC20("XXX", "XXX") {
        _grantRole(DEFAULT_ADMIN_ROLE, tx.origin);
        _grantRole(MINTER_ROLE, tx.origin);

        _mint(tx.origin, 1000 * 10 ** decimals());
```

To make sure only addresses that are in `MINTER_ROLE` can call public function `mint` use `onlyRole` modifier:
```solidity
function mint(address to, uint256 amount) public onlyRole(MINTER_ROLE) {
        _mint(to, amount);    
```

**Comment by pcaversaccio** [comment]
Please note that [OpenZeppelin Contracts version 5.0.0](https://github.com/OpenZeppelin/openzeppelin-contracts/releases/tag/v5.0.0) has made the initial `owner` explicit (see PR [#4267](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4267)).
